### PR TITLE
Log the error message when calculating basket discount.

### PR DESCRIPTION
**Pull request by vkaracic**
Let's see what fails.

### CODE REVIEW

**Comment by vkaracic** [comment]
@mjfrey

**Comment by edx-webhook** [comment]
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=1273&repo=edx%2Fecommerce"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Review by mjfrey — APPROVED** [review]
We will revert once we know why this is failing.

File: ecommerce/extensions/api/v2/views/baskets.py [hunk 1]
```diff
@@ -391,7 +391,9 @@ def get(self, request):
                     'currency': basket.currency
                 }
                 raise Exception
-        except:  # pylint: disable=bare-except
-            pass
+        except Exception as e:  # pylint: disable=bare-except
+            logger.error('Cannot calculate basket discount. Error: %s', e.message)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/baskets.py]: `logger.exception()`

**Review by clintonb — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/api/v2/views/baskets.py (hunk 1 above)
Comment (vkaracic) [reply ecommerce/extensions/api/v2/views/baskets.py]: Won't that raise the exception? I guess it doesn't matter anyway.

Comment (clintonb) [reply ecommerce/extensions/api/v2/views/baskets.py]:
```python
except:
  logger.exception('Failed to calculate basket discount for SKUs [<comma-separated list>] and voucher [<code>].')
```

Comment (clintonb) [reply ecommerce/extensions/api/v2/views/baskets.py]: Logging does not raise an exception. See https://docs.python.org/2/library/logging.html#logging.Logger.exception.

**Comment by vkaracic** [comment]
@clintonb re-review please.

**Review by clintonb — APPROVED** [review]

File: ecommerce/extensions/api/v2/views/baskets.py [hunk 2]
```diff
@@ -391,7 +391,12 @@ def get(self, request):
...
-            pass
-
+        except Exception as e:  # pylint: disable=broad-except
+            logger.exception(
+                'Failed to calculate basket discount for SKUs [%s] and voucher [%s].',
+                skus, code
+            )
+
+        if not response:
+            return Response({}, status=404)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/baskets.py]: 404 is wrong here.

**Review by clintonb — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/api/v2/views/baskets.py (hunk 2 above)
Comment (vkaracic) [reply ecommerce/extensions/api/v2/views/baskets.py]: 500?

File: ecommerce/extensions/api/v2/views/baskets.py [hunk 3]
```diff
@@ -391,7 +391,12 @@ def get(self, request):
                     'currency': basket.currency
                 }
                 raise Exception
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/baskets.py]: Why is this here?

File: ecommerce/extensions/api/v2/views/baskets.py (hunk 2 above)
Comment (clintonb) [reply ecommerce/extensions/api/v2/views/baskets.py]: Yes. Alternatively, just re-raise the exception after logging it. Django will handle the rest.

File: ecommerce/extensions/api/v2/views/baskets.py [hunk 4]
```diff
@@ -391,7 +391,12 @@ def get(self, request):
...
-        except:  # pylint: disable=bare-except
-            pass
-
+        except Exception as e:  # pylint: disable=broad-except
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/baskets.py]: No need for `as e`.

**Review by clintonb — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/api/v2/views/baskets.py (hunk 3 above)
Comment (mjfrey) [reply ecommerce/extensions/api/v2/views/baskets.py]: That is to force the transaction roll-back.

Comment (clintonb) [reply ecommerce/extensions/api/v2/views/baskets.py]: Please add a comment explaining that. We should create a custom exception. That exception can be caught and ignored. The rest can be logged. Otherwise, the logging below will just be noise in the logs.

Comment (vkaracic) [inline ecommerce/extensions/api/v2/views/baskets.py]: @mjfrey yeah why is that there? :smile:

**Comment by vkaracic** [comment]
@clintonb please re-review.

**Review by clintonb — APPROVED** [review]

File: ecommerce/extensions/api/v2/views/baskets.py [hunk 5]
```diff
@@ -390,8 +390,8 @@ def get(self, request):
                     'total_incl_tax': basket.total_incl_tax,
                     'currency': basket.currency
                 }
-                raise Exception
-        except:  # pylint: disable=bare-except
+                raise api_exceptions.TemporaryBasketException
+        except api_exceptions.TemporaryBasketException:
             pass
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/baskets.py]: There is still some value in catching unexpected exceptions, and logging the parameters that resulted in the exception being raised.

Comment (vkaracic) [reply ecommerce/extensions/api/v2/views/baskets.py]: Agreed, I've added another except-block for that purpose.

File: ecommerce/extensions/api/v2/views/baskets.py [hunk 6]
```diff
@@ -390,8 +390,13 @@ def get(self, request):
...
+        except:  # pylint: disable=bare-except
+            logger.exception(
+                'Failed to calculate basket discount for SKUs [%s] and voucher [%s].',
+                skus, code
+            )
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/baskets.py:400]: Re-`raise` the exception.

**Review by clintonb — APPROVED** [review]
